### PR TITLE
graphql-federated-graph: fix missing-Query case and upgrade path in FederatedGraphV3

### PR DESCRIPTION
**Pull request by tomhoule**
This PR should fix the end-to-end test failures due to FederatedGraphV3. It fixes two problems that lead to federated graphs without `__type` and `__schema` fields on the root query type, leading to an assertion failure in introspection.

1. The actual cause for the e2e test failures: `FederatedGraph::from_sdl()`. I assumed that we only use this with schemas produced by composition and then `FederatedGraph::render_sdl()`, but some test cases have an empty root query type. That made the provision to add `__type` and `__schema` in `from_sdl()` not trigger, and the fields were missing as a result. Thanks @obmarg for figuring this out.

To make sure that composition never produces an empty query type, the PR also contains a new composition test. We were only testing that a missing query type is rejected, we now test the empty query type case.

3. Another potential source of failures in the FederatedGraph version update logic: we weren't adding the fields when upgrading from v2 to v3. This PR contains a fix and test.

### CODE REVIEW

File: engine/crates/federated-graph/src/from_sdl.rs [hunk 1]
```diff
@@ -1014,4 +1053,180 @@ fn test_from_sdl() {
...
+          """
+          SECURITY
+
+          """
+          `EXECUTION` features provide metadata necessary for operation execution.
+          """
+          EXECUTION
+        }
+
+        type Query
```
Comment (obmarg) [inline engine/crates/federated-graph/src/from_sdl.rs:1118]: Should we actively be trying to support this?  It's not a valid schema, so doesn't seem like something we should put effort into...

Comment (tomhoule) [reply engine/crates/federated-graph/src/from_sdl.rs:1136]: No indeed, I was just writing (and pushed) a test to make it an error.

File: engine/crates/federated-graph/src/federated_graph/v3.rs [hunk 2]
```diff
@@ -208,6 +208,81 @@ impl From<super::v2::FederatedGraphV2> for FederatedGraphV3 {
...
+                    .unwrap_or_else(|| {
+                        let idx = value.strings.len();
+                        value.strings.push((*needle).to_owned());
+                        StringId(idx)
+                    })
+            });
+
+            let query_object_id = value.root_operation_types.query;
+            let original_start = {
+                let query_object_fields = object_fields[query_object_id.0].as_mut().expect("Query to have fields");
```
Comment (yoav-lavi) [inline engine/crates/federated-graph/src/federated_graph/v3.rs:230]: Is this expect here because the query must have fields at this point?

**Review by yoav-lavi — APPROVED** [review]
Looks good, see comments

File: engine/crates/composition/src/validate.rs [hunk 3]
```diff
@@ -4,9 +4,22 @@ type ValidateContext<'a> = crate::ComposeContext<'a>;
...
+    if ctx
+        .subgraphs
+        .iter_subgraphs()
+        .filter_map(|subgraph| subgraph.query_type())
+        .all(|query_type| query_type.fields().next().is_none())
+    {
+        ctx.diagnostics
+            .push_fatal(String::from("None of the subgraphs defines root query fields."));
+    }
+}
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: We have that validation in graphql-schema-validation, I think. Double checking.

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: https://github.com/grafbase/grafbase/blob/main/engine/crates/validation/tests/validation_errors/objects_nonempty.errors.txt this is the one I was thinking of, but we're not testing `type Something` without curly braces, so I'll make sure we reject that too. That code path is not taken if the schemas aren't validated before composition / in from_sdl() though. Maybe we could validate as a first step inside from_sdl(), but I'm a bit worried about breakage.

File: engine/crates/federated-graph/src/federated_graph/v3.rs (hunk 2 above)
Comment (tomhoule) [reply engine/crates/federated-graph/src/federated_graph/v3.rs:230]: yes, we added two so we are expecting at least these.

File: engine/crates/federated-graph/src/federated_graph/v1.rs [hunk 4]
```diff
@@ -1164,5 +1164,17 @@ mod tests {
...
         .unwrap();
+
+        let schema = schema.into_latest();
+        let query_object = &schema[schema.root_operation_types.query];
+
+        for field_name in ["__type", "__schema"] {
+            let name_id = schema.strings.iter().position(|s| s == field_name).unwrap();
+            assert!(
+                schema[query_object.fields.clone()].iter().any(|f| f.name.0 == name_id),
+                "Failed assertion: Query.{} must exists",
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/federated-graph/src/federated_graph/v1.rs:1175]: good catch, thanks

File: engine/crates/composition/src/validate.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: Indeed we weren't validating against `type Test` (without braces) properly, so I added that and pushed. That leaves open the question of whether we should re-validate in composition.

File: engine/crates/validation/src/validate/objects.rs [hunk 5]
```diff
@@ -13,6 +13,10 @@ pub(crate) fn validate_object<'a>(
     });
 
     ctx.with_fields(parent_type_name, &obj.fields, |ctx, fields| {
+        if fields.is_empty() {
+            diagnostics::empty_object(parent_type_name, ctx)
+        }
+
```
Comment (obmarg) [inline engine/crates/validation/src/validate/objects.rs:19]: Sorry, nitpicking this quite a lot, but should we be doing the same thing for interfaces & input objects?

File: engine/crates/composition/src/validate.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (obmarg) [reply engine/crates/composition/src/validate.rs:21]: > That leaves open the question of whether we should re-validate in composition.

Yeah....

It definitely would have made this issue less subtle, so I am for it in that sense.  What kind of breakage are you worried about?

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: That we have tests using schemas that wouldn't pass validation. But the correct thing to do would be fixing them. Let's unblock using `main` in grafbase/api and do this as a next PR.

File: engine/crates/validation/src/validate/objects.rs (hunk 5 above)
Comment (tomhoule) [reply engine/crates/validation/src/validate/objects.rs:19]: done & pushed

File: engine/crates/composition/src/validate.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (obmarg) [reply engine/crates/composition/src/validate.rs:21]: A worry I have is that `grafbase/api` still has [bad schemas in it](https://github.com/search?q=repo%3Agrafbase%2Fapi%20%22type%20Query%22&type=code) but the tests where you pulled this brach in aren't failing.

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: I did the same search locally, and I think the two `type Query`s without fields are in code branches that aren't taken in tests. I haven't dug more to see if we could remove / replace these, but I will before submitting the api PR to upgrade the deps.

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: The EMPTY_INITAL_SCHEMA looks like it's only used for diffs, we should replace that with an Option.

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: Good question, I'm perplexed now.

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: Maybe we have schemas that don't define Query at all (not even `type Query`)?

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: (checking)

[the remark this answers is not in the document]
Comment (obmarg) [reply engine/crates/composition/src/validate.rs:21]: `get_or_create_federated_graph` returns `None` for new graphs (because its only inputs are branch slug & project ID, there's nowhere for it to _get_ a federated graph from), which hits that `unwrap_or_default("type Query")` no?

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: Ok I may be confused at this point, but I think the following makes sense:

- What you said above, "I thought [this one](https://github.com/grafbase/api/blob/bea08c23c1158959b0569b9f8285182458b5688c/lambda/deployment/prepare/src/lib.rs#L173) is hit whenever you create a new federated graph, and was the cause of all the failures yesterday" is the case.
- That bad schema went through `graphql_federated_graph::from_sdl()`, causing the deployments to fail.
- With this branch, `from_sdl()` accepts schemas with empty / no Query type without crashing. That explains why the e2e tests are passing.
- If we start validating as a first step in `from_sdl()`, we should see the same tests fail again.

So it should be safe to fix the bad default schema in API, deploy the changes in this PR, then start validating. Not necessarily in this exact order, the only important part is that we start validating only after we have fixed that query. We could do it right away, but better fix main in a first step, then become strict imo.

It's a bit confusing with the different entry points so I hope this makes sense.

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: (so my comment about "I did the same search locally, and I think the two type Querys without fields are in code branches that aren't taken in tests." would be wrong in this scenario)

[the remark this answers is not in the document]
Comment (tomhoule) [reply engine/crates/composition/src/validate.rs:21]: Just saw this comment after refreshing:

>  get_or_create_federated_graph returns None for new graphs (because its only inputs are branch slug & project ID, there's nowhere for it to get a federated graph from), which hits that unwrap_or_default("type Query") no?

yes I think so

[the remark this answers is not in the document]
Comment (obmarg) [reply engine/crates/composition/src/validate.rs:21]: > So it should be safe to fix the bad default schema in API, deploy the changes in this PR, then start validating. Not necessarily in this exact order, the only important part is that we start validating only after we have fixed that query. We could do it right away, but better fix main in a first step, then become strict imo.

Yeah, cool - whatever order you want, so long as we fix all the things.
